iceoryx_hoofs/test/moduletests/test_relocatable_ptr.cpp

```diff
@@ -58,67 +99,290 @@ class RelocatablePointer_test : public Test
             std::cout << output << std::endl;
         }
     }
-    int m_fileDescriptor;
 };
 
-//@todo: extend test cases, but the relocatable is not used currently so this is postponed
-// the relocatable will be useful later, so we need to keep it
 
-// central use case of relocatable pointers
-// it is tested that moving (relocating via memcopy) the relocatable pointer causes the pointer to point to the correct
-// location relative to itself
-// this is similar to remap the shared memory (including the pointer and pointee) to a different location
-// while keeping their relative offset to each other the same
-TEST_F(RelocatablePointer_test, relocation)
+typedef ::testing::Types<int, Data, void, char*, const Data, const void> TestTypes;
+
+/// we require TYPED_TEST since we support gtest 1.8 for our safety targets
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+TYPED_TEST_CASE(Relocatable_ptr_typed_test, TestTypes);
+#pragma GCC diagnostic pop
+
+TYPED_TEST(Relocatable_ptr_typed_test, defaulCtorCreatesNullpointer)
 {
-    constexpr size_t BLOCK_SIZE = 1024;
-    // simulate 3 consecutive memory blocks on the stack
-    Memory<BLOCK_SIZE> block1;
-    Memory<BLOCK_SIZE> block2;
+    using T = typename TestFixture::DataType;
+    iox::rp::relocatable_ptr<T> rp;
+    EXPECT_EQ(rp.get(), nullptr);
+}
 
-    uint8_t* base1 = block1[0];
-    uint8_t* base2 = block2[0];
+TYPED_TEST(Relocatable_ptr_typed_test, copyCtorOfNullptrWorks)
+{
+    using T = typename TestFixture::DataType;
+    iox::rp::relocatable_ptr<T> rp1;
+    iox::rp::relocatable_ptr<T> rp2(rp1);
+    EXPECT_EQ(rp1.get(), nullptr);
+    EXPECT_EQ(rp2.get(), nullptr);
+}
 
-    int offset = BLOCK_SIZE / 2;
-    auto adr1 = reinterpret_cast<int*>(base1 + offset);
-    auto adr2 = reinterpret_cast<int*>(base2 + offset);
+TYPED_TEST(Relocatable_ptr_typed_test, moveCtorOfNullptrWorks)
+{
+    using T = typename TestFixture::DataType;
+    iox::rp::relocatable_ptr<T> rp1;
+    iox::rp::relocatable_ptr<T> rp2(std::move(rp1));
+    EXPECT_EQ(rp1.get(), nullptr);
+    EXPECT_EQ(rp2.get(), nullptr);
+}
 
-    // int write
-    *adr1 = 12;
-    *adr2 = 21;
+TYPED_TEST(Relocatable_ptr_typed_test, copyAssignmentOfNullptrWorks)
+{
+    using T = typename TestFixture::DataType;
+    // we cannot construct an actual object if T = void
+    // and it is not necessary fro most tests, we just need some non-nullptr
+    T* p = reinterpret_cast<T*>(NON_NULL_PTR);
+    iox::rp::relocatable_ptr<T> rp1;
+    iox::rp::relocatable_ptr<T> rp2(p);
+    rp2 = rp1;
+    EXPECT_EQ(rp1.get(), nullptr);
```

Comment:
It would be nice to add the automatic bool conversion operator to check if the relocatable contains a pointer or not.
```cpp
explicit operator bool() noexcept { return get() != nullptr; }
```

Then we could write code like
```cpp
if ( myRelocatable) {
  myRelocatable->dostuff = 5;
}
```

Reply:
This may already work due to the availability of implicit conversion to raw pointer (I think, have to check). I know for sure it will cause ambiguity if we add the operator bool, and I would not drop the operator T* for it (it is essential for convenience/usability).